### PR TITLE
[DO NOT MERGE] layout=intrinsic

### PR DESCRIPTION
**Pull request by jpettitt**
implements #12020 for amp-img, amp- video, amp carousel, amp-fit-text and amp-layout

### CODE REVIEW

File: builtins/amp-img.md [hunk 1]
```diff
@@ -144,6 +144,9 @@ If you want your image to scale as the window is resized but up to a maximum wid
 1. Set `layout=responsive` for `<amp-img>`.
 2. On the container of the image, specify the `max-width:<max width to display image>` CSS attribute.  Why on the container?  An `amp-img` element with `layout=responsive` is a *block-level* element, whereas, `<img>` is *inline*. Alternatively, you could set `display: inline-block` in your CSS for the amp-img element.
 
+#### The difference between responsive and intrinsic layout
+
+Both the `responsive` and `intrinsic` layouts create an image that will scale automatically.  The main difference is that the `intrinsic` layout uses an SVG image as it's scaling element.  This will make it behave in the same way as a standard html image while retaining the benifit of the browser knowing the image size on initial layout. The `intrinsic` layout will have an intrinsic size and will inflate a floated `div` until it reaches either the natural image size or a CSS constraint like `max-width`. The `responsive` layout will render 0x0 in a floated `div` becasue it takes it's size from the parent which in a float has no natural size.
```
Comment (aghassemi) [inline builtins/amp-img.md]: s\it's scaling\its scaling

/to @bpaduch for doc review

File: test/functional/test-friendly-iframe-embed.js [hunk 2]
```diff
@@ -390,56 +390,56 @@ describe('friendly-iframe-embed', () => {
 
     it('should pre-pend to html', () => {
       const html = mergeHtmlForTesting(spec);
-      expect(html).to.equal('<base href="https://acme.org/embed1"><a></a>');
+      expect(html).to.equal('<base href="https://acme.org/embed1">'
```
Comment (aghassemi) [inline test/functional/test-friendly-iframe-embed.js]: ditto, unrelated

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 3]
```diff
@@ -83,12 +83,6 @@ export const SAFEFRAME_VERSION_HEADER = 'X-AmpSafeFrameVersion';
 /** @type {string} @visibleForTesting */
 export const EXPERIMENT_FEATURE_HEADER_NAME = 'amp-ff-exps';
```
Comment (aghassemi) [inline extensions/amp-a4a/0.1/amp-a4a.js]: unrelated changes, let's revert

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js [hunk 4]
```diff
@@ -24,8 +24,6 @@ import {
   SAFEFRAME_VERSION_HEADER,
   protectFunctionWrapper,
   assignAdUrlToError,
-  EXPERIMENT_FEATURE_HEADER_NAME,
-  CSP_ENABLED_EXP_NAME,
 } from '../amp-a4a';
```
Comment (aghassemi) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: ditto, unrelated

File: extensions/amp-anim/amp-anim.md [hunk 5]
```diff
@@ -27,7 +27,7 @@ limitations under the License.
   </tr>
   <tr>
     <td class="col-fourty"><strong><a href="https://www.ampproject.org/docs/guides/responsive/control_layout.html">Supported Layouts</a></strong></td>
-    <td>fill, fixed, fixed-height, flex-item, nodisplay, responsive</td>
+    <td>fill, fixed, fixed-height, flex-item, intrinsic, nodisplay, responsive</td>
```
Comment (aghassemi) [inline extensions/amp-anim/amp-anim.md]: There are technically a lot more components that can support this given `isFixedLayout()` returns true but I do like limiting documentation for the ones that this makes sense for. I'd add `amp-video` to the current list you have as well.

File: src/friendly-iframe-embed.js [hunk 6]
```diff
@@ -280,10 +276,8 @@ function mergeHtml(spec) {
   }
 
   // Load CSP
-  if (spec.cspEnabled) {
-    result.push('<meta http-equiv=Content-Security-Policy ' +
+  result.push('<meta http-equiv=Content-Security-Policy ' +
```
Comment (aghassemi) [inline src/friendly-iframe-embed.js]: ditto: unrelated changes

File: validator/validator-main.protoascii [hunk 7]
```diff
@@ -3521,6 +3521,7 @@ tags: {  # <amp-img>
     supported_layouts: FLEX_ITEM
     supported_layouts: NODISPLAY
     supported_layouts: RESPONSIVE
+    supported_layouts: INTRINSIC
```
Comment (aghassemi) [inline validator/validator-main.protoascii]: `amp-layout` rule just below needs this too

File: validator/validator.proto [hunk 8]
```diff
@@ -238,6 +238,7 @@ message AmpLayout {
     CONTAINER = 5;
     FILL = 6;
     FLEX_ITEM = 7;
+    INTRINSIC = 8;
```
Comment (aghassemi) [inline validator/validator.proto]: let's add validator tests to `validator/testdata/feature_tests/amp_layouts.html` and  `validator/testdata/feature_tests/amp_layouts.out` for the tags that support intrinsic now

File: css/amp.css [hunk 9]
```diff
@@ -165,6 +165,17 @@ html.i-amphtml-ios-embed.i-amphtml-ios-overscroll > #i-amphtml-wrapper {
   position: relative;
 }
 
+.i-amphtml-layout-intrinsic
```
Comment (aghassemi) [inline css/amp.css]: needs an additional:
```
[layout=intrinsic][width][height]:not(.i-amphtml-layout-intrinsic),
```
if JS fails to load and `.i-amphtml-layout-intrinsic` is not applied

File: css/amp.css [hunk 10]
```diff
@@ -235,6 +246,10 @@ i-amphtml-sizer {
   right: 0;
 }
 
+.i-amphtml-layout-intrinsic .i-amphtml-sizer {
```
Comment (aghassemi) [inline css/amp.css]: looks like dead code replaced by `.i-amphtml-intrinsic-sizer`?

File: extensions/amp-fit-text/validator-amp-fit-text.protoascii [hunk 11]
```diff
@@ -38,5 +38,6 @@ tags: {  # <amp-fit-text>
     supported_layouts: FLEX_ITEM
     supported_layouts: NODISPLAY
     supported_layouts: RESPONSIVE
+    supported_layouts: INTRINSIC
```
Comment (aghassemi) [inline extensions/amp-fit-text/validator-amp-fit-text.protoascii]: /cc @Gregable @honeybadgerdontcare for validator and to ensure this does not break SSR given it is a new layout (also we need SSR support for this, see updated `applyStaticLayout` around the `layout == Layout.INTRINSIC` case)

File: src/base-element.js [hunk 12]
```diff
@@ -750,6 +750,25 @@ export class BaseElement {
...
+   *
+   * If `opt_replacedContent` is specified, it indicates whether the "replaced
+   * content" styling should be applied. Replaced content is not allowed to
+   * have its own paddings or border.
+   *
+   * @param {!Element} element
+   * @param {boolean=} opt_replacedContent
+   * @public @final
+   */
+  applyIntrinsicContent(element, opt_replacedContent) {
```
Comment (aghassemi) [inline src/base-element.js]: Do we call this from anywhere? 
We have `applyFillContent` mostly because some components like carousel need to apply it at runtime. I don't see a need for this (yet).

File: validator/validator.proto (hunk 8 above)
Comment (aghassemi) [inline validator/validator.proto]: Also there is a validator rule that enforces `width` `height` for `responsive`. Not sure if that is being applied to `intrinsic`, the test to be added above can verify that.

File: test/functional/test-layout.js [hunk 13]
```diff
@@ -275,6 +275,40 @@ describe('Layout', () => {
     expect(div.children[0].style.paddingTop).to.equal('200%');
   });
 
+
+  it('layout=intrinsic', () => {
```
Comment (aghassemi) [inline test/functional/test-layout.js]: visual tests for this would be great. please add to existing layout visual tests in `examples/visual-tests/amp-layout`

File: src/base-element.js (hunk 12 above)
Comment (jpettitt) [reply src/base-element.js]: You're right I'll remove it

File: test/functional/test-layout.js (hunk 13 above)
Comment (jpettitt) [reply test/functional/test-layout.js]: will do

File: css/amp.css (hunk 9 above)
Comment (jpettitt) [reply css/amp.css]: adding

File: css/amp.css (hunk 10 above)
Comment (jpettitt) [reply css/amp.css]: removing

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 3 above)
Comment (jpettitt) [reply extensions/amp-a4a/0.1/amp-a4a.js]: will do - there are days I hate git and days I just loath it.

File: extensions/amp-anim/amp-anim.md (hunk 5 above)
Comment (jpettitt) [reply extensions/amp-anim/amp-anim.md]: I'm going to add amp-ad to fix #12261

File: validator/validator-main.protoascii (hunk 7 above)
Comment (jpettitt) [reply validator/validator-main.protoascii]: adding

File: extensions/amp-fit-text/validator-amp-fit-text.protoascii (hunk 11 above)
Comment (honeybadgerdontcare) [reply extensions/amp-fit-text/validator-amp-fit-text.protoascii]: @jpettitt Can you file either an issue or a bug to implement instrinsic in SSR and assign to me?

Comment (honeybadgerdontcare) [reply extensions/amp-fit-text/validator-amp-fit-text.protoascii]: This shouldn't break SSR as it's designed to handle unknown layouts.

File: validator/validator-main.protoascii (hunk 7 above)
Comment (honeybadgerdontcare) [reply validator/validator-main.protoascii]: alphabetize `supported_layouts`.

**Comment by powdercloud** [comment]
https://github.com/ampproject/amphtml/blob/master/spec/amp-html-layout.md - will this need updating?

File: spec/amp-html-layout.md [hunk 14]
```diff
@@ -120,13 +120,17 @@ Supported values for the `layout` attribute:
       <td><code>flex-item</code></td>
       <td>The element and other elements in its parent with layout type <code>flex-item</code> take the parent container's remaining space when the parent is a flexible container (i.e., <code>display: flex</code>). The <code>width</code> and <code>height</code> attributes are not required.</td>
     </tr>
+    <tr>
+      <td><code>intrinsic</code></td>
+      <td>The element takes the space available to it and resizes its height automatically to the aspect ratio given by the <code>width</code> and <code>height</code> attributes. This layout works very well for most AMP elements, including <code>amp-img</code>, <code>amp-video</code>, etc.  The available space depends on the parent element and can also be customized using <code>max-width</code> CSS. The <code>width</code> and <code>height</code> attributes must be present.  This layout differs from <code>responsive</code> by having an intrinsic height and width. This is most apparent inside a floated element where a <code>responsive</code> layout will render 0x0 and <code>intrinsic</code> layout will inflate to the smaller of it's natural size or any CSS constraint. </td>
```
Comment (bpaduch) [inline spec/amp-html-layout.md]: The fragment `and can also be customized using <code>max-width</code> CSS.` does that pertain to the parent element or to the element that has the `intrinsic` layout? Please clarify wording.

File: spec/amp-html-layout.md [hunk 15]
```diff
@@ -120,13 +120,17 @@ Supported values for the `layout` attribute:
...
     </tr>
     <tr>
       <td><code>responsive</code></td>
-      <td>The element takes the space available to it and resizes its height automatically to the aspect ratio given by the <code>width</code> and <code>height</code> attributes. This layout works very well for most AMP elements, including <code>amp-img</code>, <code>amp-video</code>, etc.  The available space depends on the parent element and can also be customized using <code>max-width</code> CSS. The <code>width</code> and <code>height</code> attributes must be present.</td>
+      <td>The element takes the space available to it and resizes its height automatically to the aspect ratio given by the <code>width</code> and <code>height</code> attributes. This layout works very well for most AMP elements, including <code>amp-img</code>, <code>amp-video</code>, etc.  The available space depends on the parent element and can also be customized using <code>max-width</code> CSS. The <code>width</code> and <code>height</code> attributes must be present. This layout is similar to intrinsic but does not inflate, that is renders 0x0, when inside a floated element.</td>
```
Comment (bpaduch) [inline spec/amp-html-layout.md]: Consider rewriting as:

This layout is similar to the `intrinsic` layout; however, because the `responsive` layout has no intrinsic size, it does not inflate, that is, it renders 0x0 when inside a floated element.

File: builtins/amp-img.md [hunk 16]
```diff
@@ -144,6 +144,9 @@ If you want your image to scale as the window is resized but up to a maximum wid
 1. Set `layout=responsive` for `<amp-img>`.
 2. On the container of the image, specify the `max-width:<max width to display image>` CSS attribute.  Why on the container?  An `amp-img` element with `layout=responsive` is a *block-level* element, whereas, `<img>` is *inline*. Alternatively, you could set `display: inline-block` in your CSS for the amp-img element.
 
+#### Understanding responsive vs intrinsic layout
+
+Both the `responsive` and `intrinsic` layouts create an image that will scale automatically. The main difference is that the `intrinsic` layout uses an SVG image as it's scaling element.  This will make it behave in the same way as a standard html image while retaining the benefit of the browser knowing the image size on initial layout. The `intrinsic` layout will have an intrinsic size and will inflate a floated `div` until it reaches either the natural image size or a CSS constraint like `max-width`. The `responsive` layout will render 0x0 in a floated `div` because it takes its size from the parent which in a float has no natural size.
```
Comment (bpaduch) [inline builtins/amp-img.md]: I found this confusing.  Can you clarify:
- "uses an SVG image" - is that internal or does the developer have to specify it?
- "uses an SVG image as it's scaling element" - if `intrinsic` uses an SVG image, what does responsive use for scaling?
- when shouldn't you use `intrinsic` over `responsive` ?
- "The `intrinsic` layout will have an intrinsic size" - specified by whom, from where?  Clarify plz.
-  "CSS constraint like `max-width`." - specified on the parent yes?  Please add for clarification, otherwise the reader might imply they can use max-width on the amp-img.

**Comment by aghassemi** [comment]
@jpettitt Please ping me when it is ready for final look. Thanks.

File: builtins/amp-img.md (hunk 16 above)
Comment (jpettitt) [reply builtins/amp-img.md]: I'll take another stab at this.  The max-width can be on the element or a parent.  I think I might need some pictures and example code to make this clear.  As to when you should use intrinsic, the answer is all the time, The only reason to not make responsive behave this way is that it would be a breaking change.

**Comment by jpettitt** [comment]
Arrrgh Merge hell.  I'm closing this PR, it can't be deployed until after the holiday due to validator synchronization and holiday code freeze.  I'll make a new PR after the holiday to avoid endless rebasing.
